src/main/java/cubicchunks/worldgen/GeneratorPipeline.java

```diff
@@ -21,318 +21,82 @@
-		int numUpdated = 0;
-		while (!toUpdateSet.isEmpty()) {
-			numUpdated++;
-			GeneratorStage stage = processNext();
-			if (stage == null) {
-				skipped++;
-				continue;
-			}
-			int currentStageOrdinal = stage.getOrdinal() + 1;
-			if (currentStageOrdinal >= cubesInStage.length) {
-				finished++;
-			} else {
-				cubesInStage[currentStageOrdinal]++;
-			}
-			if (numUpdated%50 == 0) {
-				updateProgress(progress, cubesInStage, finished, skipped);
-			}
-		}
+	@Nonnull
+	public Iterator<GeneratorStage> iterator() {
+		return this.stages.iterator();
 	}
 
-	//TODO: Track progress properly
-	private void updateProgress(Progress progress, int[] cubesInStage, int finished, int skipped) {
-		int[] actualStages = new int[cubesInStage.length + 1];
-		System.arraycopy(cubesInStage, 0, actualStages, 0, cubesInStage.length);
-		actualStages[actualStages.length - 1] = finished + skipped;
-		int totalKnownUnfinished = 0;
-		//the amount of cubes that went through 0-th stage
-		totalKnownUnfinished += cubesInStage[0];
-		for (int n = 1; n < cubesInStage.length; n++) {
-			//the amount if cubes that went through n-th stage but didn't go through (n-1)-th stage.
-			//TODO: is it accurate enough?
-			totalKnownUnfinished += Math.max(0, cubesInStage[n] - cubesInStage[n - 1]);
-		}
-		int unknown = toUpdateSet.size() + finished + skipped - totalKnownUnfinished;
-		if (unknown < 0) {
-			unknown = 0;
-		}
-		//if unknown, assume 0
-		actualStages[0] += unknown;
-		for (int i = 0; i < actualStages.length; i++) {
-			int currentStage = (i == actualStages.length - 1) ?
-					actualStages[i] :
-					actualStages[i] - actualStages[i + 1];
-			if (currentStage < 0) {
-				currentStage = 0;
-			}
-			actualStages[i] = currentStage;
-		}
-		int total = 0;
-		for (int num : actualStages) {
-			total += num;
-		}
-		double totalProgress = 0;
-		for (int i = 0; i < actualStages.length; i++) {
-			double progressPart = actualStages[i]/(double) total;
-			double weightedProgress = progressPart*i/(double) (actualStages.length - 1);
-			totalProgress += weightedProgress;
-		}
-		assert totalProgress < 1.00001 && totalProgress >= 0;
-		if (totalProgress > 1) {
-			totalProgress = 1;
-		}
-		int progressInt = (int) (totalProgress*Integer.MAX_VALUE);
-		progress.setProgress(progressInt);
+	public int stageCount() {
+		// The stage LIVE does not count as a proper stage.
+		return this.stages.size() - 1;
 	}
 
-	@Nullable public GeneratorStage getStage(String name) {
+	@Nullable
+	public GeneratorStage getStage(@Nonnull String name) {
 		return this.stageMap.get(name);
 	}
 
+	@Nonnull
 	public GeneratorStage getFirstStage() {
 		return this.stages.get(0);
 	}
 
-	public void remove(CubeCoords coords) {
-		Cube cube = this.cubeProvider.getCube(coords);
-		if(cube != null) {
-			//removing from array list is a bit expensive, remove from set and skip cubes that don't exist in the set
-			this.toUpdateSet.remove(coords);
-			this.dependentCubeManager.unregister(cube);
-		}
-	}
```

Comment:
Note: comment to changes in the whole class

So GeneratorPipeline is now more like stage registry, right?